File: src/main/java/com/tibiawiki/domain/objects/Corpse.java
package com.tibiawiki.domain.objects;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.tibiawiki.domain.enums.Article;
import com.tibiawiki.domain.enums.Status;
import com.tibiawiki.domain.enums.YesNo;
import lombok.Builder;
import lombok.Getter;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;

@Getter
@Component
public class Corpse extends WikiObject {

    private final String flavortext;
    private final YesNo skinable;
    private final String product;
    private final String liquid;
    private final Integer stages;
    private final String firstDecaytime; // FIXME should be Seconds
    private final String secondDecaytime; // FIXME should be Seconds
    private final String thirdDecaytime; // FIXME should be Seconds
    private final Integer firstVolume;
    private final Integer secondVolume;
    private final Integer thirdVolume;
    private final BigDecimal firstWeight;
    private final BigDecimal secondWeight;
    private final BigDecimal thirdWeight;
    private final String corpseof;
    private final String sellto;

    private Corpse() {
        this.flavortext = null;
        this.skinable = null;
        this.product = null;
        this.liquid = null;
        this.stages = null;
        this.firstDecaytime = null;
        this.secondDecaytime = null;
        this.thirdDecaytime = null;
        this.firstVolume = null;
        this.secondVolume = null;
        this.thirdVolume = null;
        this.firstWeight = null;
        this.secondWeight = null;
        this.thirdWeight = null;
        this.corpseof = null;
        this.sellto = null;
    }

    @Builder
    private Corpse(String name, Article article, String actualname, String implemented, String notes, String history,
                   Status status, String flavortext, YesNo skinable, String product, String liquid, Integer stages,
                   String firstDecaytime, String secondDecaytime, String thirdDecaytime, Integer firstVolume,
                   Integer secondVolume, Integer thirdVolume, BigDecimal firstWeight, BigDecimal secondWeight, BigDecimal thirdWeight,
                   String corpseof, String sellto) {
        super(name, article, actualname, null, implemented, notes, history, status);
        this.flavortext = flavortext;
        this.skinable = skinable;
        this.product = product;
        this.liquid = liquid;
        this.stages = stages;
        this.firstDecaytime = firstDecaytime;
        this.secondDecaytime = secondDecaytime;
        this.thirdDecaytime = thirdDecaytime;
        this.firstVolume = firstVolume;
        this.secondVolume = secondVolume;
        this.thirdVolume = thirdVolume;
        this.firstWeight = firstWeight;
        this.secondWeight = secondWeight;
        this.thirdWeight = thirdWeight;
        this.corpseof = corpseof;
        this.sellto = sellto;
    }

    @JsonGetter("1decaytime")
    public String getFirstDecaytime() {
        return firstDecaytime;
    }

    @JsonGetter("2decaytime")
    public String getSecondDecaytime() {
        return secondDecaytime;
    }

    @JsonGetter("3decaytime")
    public String getThirdDecaytime() {
        return thirdDecaytime;
    }

    @JsonGetter("1volume")
    public Integer getFirstVolume() {
        return firstVolume;
    }

    @JsonGetter("2volume")
    public Integer getSecondVolume() {
        return secondVolume;
    }

    @JsonGetter("3volume")
    public Integer getThirdVolume() {
        return thirdVolume;
    }

    @JsonGetter("1weight")
    public BigDecimal getFirstWeight() {
        return firstWeight;
    }

    @JsonGetter("2weight")
    public BigDecimal getSecondWeight() {
        return secondWeight;
    }

    @JsonGetter("3weight")
    public BigDecimal getThirdWeight() {
        return thirdWeight;
    }

    @Override
    public List<String> fieldOrder() {
        return Arrays.asList("name", "article", "actualname", "flavortext", "skinable", "product", "liquid", "stages",
                "1decaytime", "2decaytime", "3decaytime", "1volume", "2volume", "3volume", "1weight", "2weight",
                "3weight", "corpseof", "sellto", "notes", "implemented", "history", "status");
    }
}

File: src/main/kotlin/com/tibiawiki/domain/objects/Corpse.kt
package com.tibiawiki.domain.objects

import com.fasterxml.jackson.annotation.JsonProperty
import com.tibiawiki.domain.enums.Article
import com.tibiawiki.domain.enums.InfoboxTemplate
import com.tibiawiki.domain.enums.Status
import com.tibiawiki.domain.enums.YesNo
import java.math.BigDecimal

data class Corpse(
    override val name: String? = null,
    override val article: Article? = null,
    override val actualname: String? = null,
    override val implemented: String? = null,
    override val notes: String? = null,
    override val history: String? = null,
    override val status: Status? = null,
    val flavortext: String? = null,
    val skinable: YesNo? = null,
    val product: String? = null,
    val liquid: String? = null,
    val stages: Int? = null,
    @JsonProperty("1decaytime") val firstDecaytime: String? = null,
    @JsonProperty("2decaytime") val secondDecaytime: String? = null,
    @JsonProperty("3decaytime") val thirdDecaytime: String? = null,
    @JsonProperty("1volume") val firstVolume: Int? = null,
    @JsonProperty("2volume") val secondVolume: Int? = null,
    @JsonProperty("3volume") val thirdVolume: Int? = null,
    @JsonProperty("1weight") val firstWeight: BigDecimal? = null,
    @JsonProperty("2weight") val secondWeight: BigDecimal? = null,
    @JsonProperty("3weight") val thirdWeight: BigDecimal? = null,
    val corpseof: String? = null,
    val sellto: String? = null
) : WikiObject(
    name = name,
    article = article,
    actualname = actualname,
    implemented = implemented,
    notes = notes,
    history = history,
    status = status
) {
    override fun getTemplateType(): String {
        return InfoboxTemplate.CORPSE.templateName
    }

    override fun fieldOrder(): List<String> {
        return listOf(
            "name", "article", "actualname", "flavortext", "skinable", "product", "liquid", "stages",
            "1decaytime", "2decaytime", "3decaytime", "1volume", "2volume", "3volume", "1weight", "2weight",
            "3weight", "corpseof", "sellto", "notes", "implemented", "history", "status"
        )
    }
}
